### PR TITLE
Applesoft compiler fixes

### PR DESCRIPTION
**Pull request by iflan**
These changes fix several bugs in the Applesoft compiler, including:

* Lowercase letters are preserved in strings, comments, and `DATA` statements.
* Lines are always sorted no matter what order they are entered.
* Correctly prefers `ATN` to `AT`.
* Correctly prefers `A TO` to `AT O`.

These changes add tests for the compiler as well.

### CODE REVIEW

File: js/applesoft/tokens.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
+
+/** Map from token to keyword */
+export const TOKEN_TO_STRING = {
```
Comment (whscullin) [inline js/applesoft/tokens.ts:3]: Typing these as `Record<byte, string>` and `Record<string, byte` will avoid having to cast later on.

File: test/js/applesoft/compiler.spec.ts [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+        expect(ram.read(0x08, 0x0e)).toBe(0x57); // W
+        expect(ram.read(0x08, 0x0f)).toBe(0x4F); // O
+        expect(ram.read(0x08, 0x10)).toBe(0x52); // R
+        expect(ram.read(0x08, 0x11)).toBe(0x4C); // L
+        expect(ram.read(0x08, 0x12)).toBe(0x44); // D
+        expect(ram.read(0x08, 0x13)).toBe(0x21); // !
+        expect(ram.read(0x08, 0x14)).toBe(0x22); // "
+        expect(ram.read(0x08, 0x15)).toBe(0x00); // end of line
+        expect(ram.read(0x08, 0x16)).toBe(0x00); // end of program low
+        expect(ram.read(0x08, 0x17)).toBe(0x00); // end of program high
```
Comment (whscullin) [inline test/js/applesoft/compiler.spec.ts:67]: I wonder if using `TestMemory` and adding some array accessors would make these a less dense.

File: test/js/applesoft/compiler.spec.ts [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+// Zero page locations used by Applesoft. The names come from
+// the commented decompilation produced by the Merlin Pro
+// assembler, revision 4/27/84. There is evidence from
+// https://www.pagetable.com/?p=774 that the original Microsoft
+// BASIC source code used these names as well.
+const TXTTAB = 0x67; // start of program, word
+const VARTAB = 0x69; // start of variables, word
+const ARYTAB = 0x6B; // start of arrays, word
+const STREND = 0x6D; // end of strings, word
+const PRGEND = 0xAF; // end of program, word
```
Comment (whscullin) [inline test/js/applesoft/compiler.spec.ts:14]: Most of these get reused elsewhere, maybe export from here or a constants file in this directory?

**Review by whscullin — COMMENTED** [review]
Awesome to get some tests around this. I made a couple suggestions and it needs a linter pass.

File: test/js/applesoft/compiler.spec.ts (hunk 2 above)
Comment (iflan) [reply test/js/applesoft/compiler.spec.ts:67]: This is a good point. I've been trying to figure out something reasonable, but the best I get is:

```typescript
expect(ram.range(0x0801, 0x0817)).toEqual([
    0x16, // pointer to next line low
    0x08, // pointer to next line high
    ...
]);
```

Another choice could be:

```typescript
expect(ram.range(0x0801, 0x0817)).toMatch({
    0x0801: 0x16, // pointer to next line low
    0x0801: 0x08, // pointer to next line high
    ...
});
```

Thoughts?

File: js/applesoft/tokens.ts (hunk 1 above)
Comment (iflan) [reply js/applesoft/tokens.ts]: I was trying to treat the map as a type, hence the `as const` at the end. I don't think that extra typing is buying anything, so I'll remove it and see what happens.

File: test/js/applesoft/compiler.spec.ts (hunk 2 above)
Comment (whscullin) [reply test/js/applesoft/compiler.spec.ts:67]: Sorry I completely spaced on this - if you loop back to this I like the first version. Using the constants for tokens might also help readability.

**Pull request merged**